Import TF-M NS APIs into Mbed OS

urutva opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->
After successfully building a TF-M image for a particular target or all
supported targets import TF-M APIs into Mbed OS. These NS APIs must be
used by Mbed OS applications to request secure services from TF-M.

Signed-off-by: Devaraj Ranganna <devaraj.ranganna@arm.com>
<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->

#### Impact of changes <!-- Optional -->
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

#### Migration actions required <!-- Optional -->
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

### Documentation <!-- Required -->

<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [x] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [x] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

<!--
    Request additional reviewers with @username or @team
-->
@Patater @jainvikas8 

----------------------------------------------------------------------------------------------------------------

urutva committed:
Import TF-M NS APIs into Mbed OS

After successfully building a TF-M image for a particular target or all
supported targets import TF-M APIs into Mbed OS. These NS APIs must be
used by Mbed OS applications to request secure services from TF-M.

Signed-off-by: Devaraj Ranganna <devaraj.ranganna@arm.com>

Patater commented on tools/psa/build_tfm.py:
```diff
@@ -419,6 +420,37 @@ def _copy_binaries(source, destination, toolchain, target):
+            try:
+                shutil.copy2(src_file, dst_file)
+            except FileNotFoundError:
+                # Wordaround to handle app/os_wrapper_cmsis_rtos_v2.c located
```
```diff
- Wordaround
+ Workaround
```

Patater commented on tools/psa/build_tfm.py:
```diff
@@ -419,6 +420,37 @@ def _copy_binaries(source, destination, toolchain, target):
+                shutil.copy2(src_file, dst_file)
+            except FileNotFoundError:
+                # Wordaround to handle app/os_wrapper_cmsis_rtos_v2.c located
+                # in the root directory of TF-M
```
Could you explain this comment more to explain what the workaround is and why it is needed? It's not clear why this file can't be handled in a manner consistent with the other files.

Why does a file sourced from `install/export/` get treated differently from a file sourced from `app/`?

urutva replied:
Updated the comment with more details.

Patater commented on tools/psa/tfm_ns_import.json:
```diff
@@ -0,0 +1,66 @@
+            "dst_folder": "features/FEATURE_PSA/TARGET_TFM/include/psa_manifest"
+        }
+    ]
+}
```
Suggest to add a newline at the end of the file.

urutva replied:
Fixed

Patater reviewed (changes requested):
Just to check, the files added here are from the TF-M version already in `features/FEATURE_PSA/TARGET_TFM/VERSION.txt`, and this is why we don't need to update `VERSION.txt` in this PR. Right?

Patater commented on tools/psa/build_tfm.py:
```diff
@@ -419,6 +420,41 @@ def _copy_binaries(source, destination, toolchain, target):
+                # list of files and folder relative to cmake build folder. The
+                # only exception is the OS abstraction layer app/os_wrapper_cmsis_rtos_v2.c
+                # which is not exported by TF-M. Therefore it is handled via an
+                # exception
```
I recommend adding a full stop to the end of complete sentences. I'd also recommend replacing "via" with "as" in this instance, as well as another comma for readability.

```diff
- Therefore it is handled via an exception
+ Therefore, it is handled as an exception.
```

0xc0170 replied:
@devran01 Do you want to update this, we can run CI

urutva replied:
Fixed

Patater reviewed (approved):
Minor English improvement recommendation

Otherwise, LGTM

urutva committed:
Import missing files

TF-M dual cpu build doesn't export following files tfm_multi_core_api.c,
tfm_multi_core_psa_ns_api.c, tfm_ns_mailbox.c, tfm_mailbox.h,
tfm_multi_core_api.h, tfm_ns_mailbox.h which are required to
build dual cpu images.

Signed-off-by: Devaraj Ranganna <devaraj.ranganna@arm.com>

urutva committed:
Fix `extra_labels` issue in `PSA_DUAL_CORE_NSPE`

The `extra_labels` attribute in `PSA_DUAL_CORE_NSPE` contained incorrect
label `TARGET_TFM_TWINCPU` which should have been `TFM_TWINCPU`.

Signed-off-by: Devaraj Ranganna <devaraj.ranganna@arm.com>
